### DISCUSSION TITLE
Does Infinispan TransactionManagerLookup support WebSphere Liberty?

### DISCUSSION THREAD

**Discussion by RangaJ96** [infinispan/infinispan#12052] [Q&A] [answered]
I was trying to deploy a Java EE application within a WebSphere Liberty BASE server in a Windows environment, within this application I'm using Infinispan as my transactional cache. I faced the following issue when the application tried to access the transactional cache.

`java.lang.IllegalArgumentException: Cannot create a transactional context without a valid Transaction instance.
    at org.infinispan.context.impl.TransactionalInvocationContextFactory.createInvocationContext(TransactionalInvocationContextFactory.java:63)
    at org.infinispan.context.impl.TransactionalInvocationContextFactory.createInvocationContext(TransactionalInvocationContextFactory.java:57)
    at org.infinispan.cache.impl.CacheImpl.lock(CacheImpl.java:995)
    at org.infinispan.cache.impl.CacheImpl.lock(CacheImpl.java:980)
    at org.infinispan.cache.impl.EncoderCache.lock(EncoderCache.java:307)`

Below are the features I have used and my server.xml looks something like the below,
![image](https://github.com/infinispan/infinispan/assets/49330699/d913edbb-734b-4729-bdba-411f009a36cb)

Also, this is the way I configured the Infinispan transactional cache template.

`<local-cache-configuration name="local-txn-cache-template">
            <locking isolation="READ_COMMITTED" acquire-timeout="15000" concurrency-level="1"></locking>
            <transaction mode="NON_XA" locking="PESSIMISTIC"
                         transaction-manager-lookup="org.infinispan.transaction.lookup.GenericTransactionManagerLookup"
                         complete-timeout="15000"/>
        </local-cache-configuration>`

Other than that my application getting the following failures in the Infinispan transaction manager lookup stage.

![image](https://github.com/infinispan/infinispan/assets/49330699/64e48ace-6906-453b-a781-311e1d5e5747)

Can anyone verify whether Infinispan supports transactional caching with WebSphere Liberty? Thank you

**Comment by pruivo** [comment]
It looks like the `TransactionManager` in Liberty needs to be fetched from `com.ibm.tx.jta.TransactionManagerFactory`. Going to add it to the list of known transaction managers. 
In the meantime, you may create a `TransactionManagerLookup` implementation to fetch the `TransactionManager` from the correct class. Docs are here: https://infinispan.org/docs/stable/titles/configuring/configuring.html#transaction_manager and an example here: https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/transaction/lookup/WildflyTransactionManagerLookup.java
I hope this helps.

**Reply by RangaJ96** [reply]
Dear Pruivo, thanks for your help, may I know if is it okay to add the above factory to the TransactionManagerFactory in LookupNames class as below,

![image](https://github.com/infinispan/infinispan/assets/49330699/16606b83-a11d-49a1-b010-feec1926a6db)

**Reply by RangaJ96** [reply, marked as answer]
@pruivo I applied the above change to the Infinispan and now WebSphere Liberty is able to lookup the transaction manager without any issue, my concern is also solved now. thanks a lot @pruivo.

<img width="656" alt="image" src="https://github.com/infinispan/infinispan/assets/49330699/c1087cad-51c2-428b-a411-75d581490de8">

**Reply by pruivo** [reply]
Thank you for verifying it @RangaJ96 
I'm going to add it to Infinispan 14 and Infinispan 15. Tracking JIRA: https://issues.redhat.com/browse/ISPN-15755

**Reply by RangaJ96** [reply]
It's my pleasure to hear that, thanks for your kind assistance @pruivo
